tty/pc-tty/ttypc_kbd.c

```diff
@@ -14,300 +14,388 @@
+			case KB_CTL:
+				ttypc->shiftst |= KB_CTL;
+				break;
+
+			case KB_SHIFT:
+				ttypc->shiftst |= KB_SHIFT;
+				break;
+
+			case KB_ALT:
+				if (ext)
+					ttypc->shiftst |= KB_ALTGR;
+				else
+					ttypc->shiftst |= KB_ALT;
+				break;
+
+			/* Function keys */
+			case KB_FUNC:
+			/* Regular ASCII */
+			case KB_ASCII:
+				/* Keys with extended scan codes don't depend on any modifiers */
+				if (ext) {
+					/* Handles keypad '/' key */
+					s = scodes[dt].unshift;
+					break;
+				}
+				/* Control modifier */
+				if (ttypc->shiftst & KB_CTL) {
+					s = scodes[dt].ctl;
+					break;
+				}
+
+				/* Right alt and right alt with shift modifiers */
+				if (ttypc->shiftst & KB_ALTGR) {
+					if (ttypc->shiftst & KB_SHIFT)
+						s = scodes[dt].shift_altgr;
+					else
+						s = scodes[dt].altgr;
+				}
+				/* Shift modifier */
+				else if (ttypc->shiftst & KB_SHIFT) {
+					s = scodes[dt].shift;
+				}
+				/* No modifiers */
+				else {
+					s = scodes[dt].unshift;
+				}
+
+				/* Caps lock */
+				if ((ttypc->lockst & KB_CAPS) && (*scodes[dt].unshift >= 'a') && (*scodes[dt].unshift <= 'z')) {
+					if (s == scodes[dt].altgr)
+						s = scodes[dt].shift_altgr;
+					else if (s == scodes[dt].shift_altgr)
+						s = scodes[dt].altgr;
+					else if (s == scodes[dt].shift)
+						s = scodes[dt].unshift;
+					else if (s == scodes[dt].unshift)
+						s = scodes[dt].shift;
+				}
+				break;
+
+			/* Keys without meaning */
+			case KB_NONE:
+				break;
 
+			/* Keypad */
+			case KB_KP:
+				/* Keys with extended scan codes don't depend on any modifiers */
+				if (ext) {
+					/* Handles DEL, HOME, END, PU, PD and arrow (non keypad) keys */
+					s = scodes[dt].shift;
+					break;
+				}
+
+				/* Shift modifier */
 				if (ttypc->shiftst & KB_SHIFT)
-					more = scan_codes[dt].unshift;
+					s = scodes[dt].shift;
+				/* Control modifier */
+				else if (ttypc->shiftst & KB_CTL)
+					s = scodes[dt].ctl;
+				/* No modifiers */
 				else
-					more = scan_codes[dt].shift;
+					s = scodes[dt].unshift;
+
+				/* Num lock */
+				if (ttypc->lockst & KB_NUM) {
+					if (s == scodes[dt].shift)
+						s = scodes[dt].unshift;
+					else if ((s == scodes[dt].ctl) || (s == scodes[dt].unshift))
+						s = scodes[dt].shift;
+				}
+				break;
 			}
+		}
 
-			ttypc->extended = 0;
-			break;
-		
-		/* Key without meaning */
-		case KB_NONE:
-			break;
-			
-		/* Function key */
-		case KB_FUNC: {
-			if (ttypc->shiftst & KB_SHIFT)
-				more = scan_codes[dt].shift;
-			else if (ttypc->shiftst & KB_CTL)
-				more = scan_codes[dt].ctl;
-			else
-				more = scan_codes[dt].unshift;
-			
-			ttypc->extended = 0;
-			break;
+		/* AutoRepeat Mode */
+		if (!ttypc->vt->arm && (dt == lkey) && (ext == lext))
+			s = NULL;
+
+		/* Reboot sequence - Ctl-Alt-Del */
+		if (s && ext && !strcmp(s, "\033[3~") && (ttypc->shiftst == (KB_CTL | KB_ALT))) {
+			if (reboot(PHOENIX_REBOOT_MAGIC) < 0)
+				ttypc_vt_write(ttypc->vt, 0, erreboot, sizeof(erreboot));
```

Comment:
sizeof(erreboot) = 4 (sizeof(char *)) It should be:
`static const char erreboot[] = "Failed to reboot\n";`
and `ttypc_vt_write ` should expect pointer to const.

Reply:
Fixed.